### PR TITLE
Add admin prometheus metrics

### PR DESCRIPTION
**Pull request by sinhaashish**
Add a new command  `mc admin prometheus metrics <alias>`

```
ashish@ashish:~/code/go/src/github/minio/mc$ ./mc support metrics myminio
# HELP minio_cluster_capacity_raw_free_bytes Total free capacity online in the cluster.
# TYPE minio_cluster_capacity_raw_free_bytes gauge
minio_cluster_capacity_raw_free_bytes{server="127.0.0.1:9000"} 1.437033086976e+12
# HELP minio_cluster_capacity_raw_total_bytes Total capacity online in the cluster.
# TYPE minio_cluster_capacity_raw_total_bytes gauge
minio_cluster_capacity_raw_total_bytes{server="127.0.0.1:9000"} 1.90748696576e+12
# HELP minio_cluster_capacity_usable_free_bytes Total free usable capacity online in the cluster.
# TYPE minio_cluster_capacity_usable_free_bytes gauge
minio_cluster_capacity_usable_free_bytes{server="127.0.0.1:9000"} 7.18516543488e+11
# HELP minio_cluster_capacity_usable_total_bytes Total usable capacity online in the cluster.
# TYPE minio_cluster_capacity_usable_total_bytes gauge
minio_cluster_capacity_usable_total_bytes{server="127.0.0.1:9000"} 9.5374348288e+11
# HELP minio_cluster_disk_free_inodes Total free inodes.
# TYPE minio_cluster_disk_free_inodes gauge
minio_cluster_disk_free_inodes{disk="/tmp/1",server="127.0.0.1:9000"} 3.0160316e+07
minio_cluster_disk_free_inodes{disk="/tmp/2",server="127.0.0.1:9000"} 3.0160316e+07
minio_cluster_disk_free_inodes{disk="/tmp/3",server="127.0.0.1:9000"} 3.0160316e+07
minio_cluster_disk_free_inodes{disk="/tmp/4",server="127.0.0.1:9000"} 3.0160316e+07
# HELP minio_cluster_disk_offline_total Total disks offline.
# TYPE minio_cluster_disk_offline_total gauge
minio_cluster_disk_offline_total{server="127.0.0.1:9000"} 0
# HELP minio_cluster_disk_online_total Total disks online.
# TYPE minio_cluster_disk_online_total gauge
minio_cluster_disk_online_total{server="127.0.0.1:9000"} 4
# HELP minio_cluster_disk_total Total disks.
# TYPE minio_cluster_disk_total gauge
minio_cluster_disk_total{server="127.0.0.1:9000"} 4
# HELP minio_cluster_nodes_offline_total Total number of MinIO nodes offline.
# TYPE minio_cluster_nodes_offline_total gauge
minio_cluster_nodes_offline_total{server="127.0.0.1:9000"} 0
# HELP minio_cluster_nodes_online_total Total number of MinIO nodes online.
# TYPE minio_cluster_nodes_online_total gauge
minio_cluster_nodes_online_total{server="127.0.0.1:9000"} 1
# HELP minio_heal_objects_heal_total Objects healed in current self healing run
# TYPE minio_heal_objects_heal_total gauge
minio_heal_objects_heal_total{server="127.0.0.1:9000",type="bucket"} 1
minio_heal_objects_heal_total{server="127.0.0.1:9000",type="object"} 2
# HELP minio_heal_objects_total Objects scanned in current self healing run
# TYPE minio_heal_objects_total gauge
minio_heal_objects_total{server="127.0.0.1:9000",type="bucket"} 1
minio_heal_objects_total{server="127.0.0.1:9000",type="bucket-metadata"} 2
# HELP minio_heal_time_last_activity_nano_seconds Time elapsed (in nano seconds) since last self healing activity. This is set to -1 until initial self heal activity
# TYPE minio_heal_time_last_activity_nano_seconds gauge
minio_heal_time_last_activity_nano_seconds{server="127.0.0.1:9000"} 1.82793979077e+11
# HELP minio_inter_node_traffic_errors_total Total number of failed internode calls.
# TYPE minio_inter_node_traffic_errors_total counter
minio_inter_node_traffic_errors_total{server="127.0.0.1:9000"} 0
# HELP minio_inter_node_traffic_received_bytes Total number of bytes received from other peer nodes.
# TYPE minio_inter_node_traffic_received_bytes counter
minio_inter_node_traffic_received_bytes{server="127.0.0.1:9000"} 0
# HELP minio_inter_node_traffic_sent_bytes Total number of bytes sent to the other peer nodes.
# TYPE minio_inter_node_traffic_sent_bytes counter
minio_inter_node_traffic_sent_bytes{server="127.0.0.1:9000"} 2049
# HELP minio_node_disk_free_bytes Total storage available on a disk.
# TYPE minio_node_disk_free_bytes gauge
minio_node_disk_free_bytes{disk="/tmp/1",server="127.0.0.1:9000"} 3.59258271744e+11
minio_node_disk_free_bytes{disk="/tmp/2",server="127.0.0.1:9000"} 3.59258271744e+11
minio_node_disk_free_bytes{disk="/tmp/3",server="127.0.0.1:9000"} 3.59258271744e+11
minio_node_disk_free_bytes{disk="/tmp/4",server="127.0.0.1:9000"} 3.59258271744e+11
# HELP minio_node_disk_total_bytes Total storage on a disk.
# TYPE minio_node_disk_total_bytes gauge
minio_node_disk_total_bytes{disk="/tmp/1",server="127.0.0.1:9000"} 4.7687174144e+11
minio_node_disk_total_bytes{disk="/tmp/2",server="127.0.0.1:9000"} 4.7687174144e+11
minio_node_disk_total_bytes{disk="/tmp/3",server="127.0.0.1:9000"} 4.7687174144e+11
minio_node_disk_total_bytes{disk="/tmp/4",server="127.0.0.1:9000"} 4.7687174144e+11
# HELP minio_node_disk_used_bytes Total storage used on a disk.
# TYPE minio_node_disk_used_bytes gauge
minio_node_disk_used_bytes{disk="/tmp/1",server="127.0.0.1:9000"} 1.17613469696e+11
minio_node_disk_used_bytes{disk="/tmp/2",server="127.0.0.1:9000"} 1.17613469696e+11
minio_node_disk_used_bytes{disk="/tmp/3",server="127.0.0.1:9000"} 1.17613469696e+11
minio_node_disk_used_bytes{disk="/tmp/4",server="127.0.0.1:9000"} 1.17613469696e+11
# HELP minio_node_file_descriptor_limit_total Limit on total number of open file descriptors for the MinIO Server process.
# TYPE minio_node_file_descriptor_limit_total gauge
minio_node_file_descriptor_limit_total{server="127.0.0.1:9000"} 1.048576e+06
# HELP minio_node_file_descriptor_open_total Total number of open file descriptors by the MinIO Server process.
# TYPE minio_node_file_descriptor_open_total gauge
minio_node_file_descriptor_open_total{server="127.0.0.1:9000"} 14
# HELP minio_node_go_routine_total Total number of go routines running.
# TYPE minio_node_go_routine_total gauge
minio_node_go_routine_total{server="127.0.0.1:9000"} 414
# HELP minio_node_ilm_expiry_pending_tasks Number of pending ILM expiry tasks in the queue.
# TYPE minio_node_ilm_expiry_pending_tasks gauge
minio_node_ilm_expiry_pending_tasks{server="127.0.0.1:9000"} 0
# HELP minio_node_ilm_transition_active_tasks Number of active ILM transition tasks.
# TYPE minio_node_ilm_transition_active_tasks gauge
minio_node_ilm_transition_active_tasks{server="127.0.0.1:9000"} 0
# HELP minio_node_ilm_transition_pending_tasks Number of pending ILM transition tasks in the queue.
# TYPE minio_node_ilm_transition_pending_tasks gauge
minio_node_ilm_transition_pending_tasks{server="127.0.0.1:9000"} 0
# HELP minio_node_ilm_versions_scanned Total number of object versions checked for ilm actions since server start
# TYPE minio_node_ilm_versions_scanned counter
minio_node_ilm_versions_scanned{server="127.0.0.1:9000"} 0
# HELP minio_node_io_rchar_bytes Total bytes read by the process from the underlying storage system including cache, /proc/[pid]/io rchar
# TYPE minio_node_io_rchar_bytes counter
minio_node_io_rchar_bytes{server="127.0.0.1:9000"} 9.038556e+06
# HELP minio_node_io_read_bytes Total bytes read by the process from the underlying storage system, /proc/[pid]/io read_bytes
# TYPE minio_node_io_read_bytes counter
minio_node_io_read_bytes{server="127.0.0.1:9000"} 532480
# HELP minio_node_io_wchar_bytes Total bytes written by the process to the underlying storage system including page cache, /proc/[pid]/io wchar
# TYPE minio_node_io_wchar_bytes counter
minio_node_io_wchar_bytes{server="127.0.0.1:9000"} 2.4465097e+07
# HELP minio_node_io_write_bytes Total bytes written by the process to the underlying storage system, /proc/[pid]/io write_bytes
# TYPE minio_node_io_write_bytes counter
minio_node_io_write_bytes{server="127.0.0.1:9000"} 2.4576e+07
# HELP minio_node_process_cpu_total_seconds Total user and system CPU time spent in seconds.
# TYPE minio_node_process_cpu_total_seconds counter
minio_node_process_cpu_total_seconds{server="127.0.0.1:9000"} 0.96
# HELP minio_node_process_resident_memory_bytes Resident memory size in bytes.
# TYPE minio_node_process_resident_memory_bytes gauge
minio_node_process_resident_memory_bytes{server="127.0.0.1:9000"} 1.91950848e+08
# HELP minio_node_process_starttime_seconds Start time for MinIO process per node, time in seconds since Unix epoc.
# TYPE minio_node_process_starttime_seconds gauge
minio_node_process_starttime_seconds{server="127.0.0.1:9000"} 1.64985653326e+09
# HELP minio_node_process_uptime_seconds Uptime for MinIO process per node in seconds.
# TYPE minio_node_process_uptime_seconds gauge
minio_node_process_uptime_seconds{server="127.0.0.1:9000"} 182.894176302
# HELP minio_node_scanner_bucket_scans_finished Total number of bucket scans finished since server start
# TYPE minio_node_scanner_bucket_scans_finished counter
minio_node_scanner_bucket_scans_finished{server="127.0.0.1:9000"} 0
# HELP minio_node_scanner_bucket_scans_started Total number of bucket scans started since server start
# TYPE minio_node_scanner_bucket_scans_started counter
minio_node_scanner_bucket_scans_started{server="127.0.0.1:9000"} 0
# HELP minio_node_scanner_directories_scanned Total number of directories scanned since server start.
# TYPE minio_node_scanner_directories_scanned counter
minio_node_scanner_directories_scanned{server="127.0.0.1:9000"} 0
# HELP minio_node_scanner_objects_scanned Total number of unique objects scanned since server start.
# TYPE minio_node_scanner_objects_scanned counter
minio_node_scanner_objects_scanned{server="127.0.0.1:9000"} 0
# HELP minio_node_scanner_versions_scanned Total number of object versions scanned since server start.
# TYPE minio_node_scanner_versions_scanned counter
minio_node_scanner_versions_scanned{server="127.0.0.1:9000"} 0
# HELP minio_node_syscall_read_total Total read SysCalls to the kernel. /proc/[pid]/io syscr
# TYPE minio_node_syscall_read_total counter
minio_node_syscall_read_total{server="127.0.0.1:9000"} 959
# HELP minio_node_syscall_write_total Total write SysCalls to the kernel. /proc/[pid]/io syscw
# TYPE minio_node_syscall_write_total counter
minio_node_syscall_write_total{server="127.0.0.1:9000"} 206
# HELP minio_s3_requests_incoming_total Volatile number of total incoming S3 requests
# TYPE minio_s3_requests_incoming_total gauge
minio_s3_requests_incoming_total{server="127.0.0.1:9000"} 0
# HELP minio_s3_requests_rejected_auth_total Total number S3 requests rejected for auth failure.
# TYPE minio_s3_requests_rejected_auth_total counter
minio_s3_requests_rejected_auth_total{server="127.0.0.1:9000"} 0
# HELP minio_s3_requests_rejected_header_total Total number S3 requests rejected for invalid header.
# TYPE minio_s3_requests_rejected_header_total counter
minio_s3_requests_rejected_header_total{server="127.0.0.1:9000"} 0
# HELP minio_s3_requests_rejected_invalid_total Total number S3 invalid requests.
# TYPE minio_s3_requests_rejected_invalid_total counter
minio_s3_requests_rejected_invalid_total{server="127.0.0.1:9000"} 0
# HELP minio_s3_requests_rejected_timestamp_total Total number S3 requests rejected for invalid timestamp.
# TYPE minio_s3_requests_rejected_timestamp_total counter
minio_s3_requests_rejected_timestamp_total{server="127.0.0.1:9000"} 0
# HELP minio_s3_requests_waiting_total Number of S3 requests in the waiting queue
# TYPE minio_s3_requests_waiting_total gauge
minio_s3_requests_waiting_total{server="127.0.0.1:9000"} 0
# HELP minio_s3_traffic_received_bytes Total number of s3 bytes received.
# TYPE minio_s3_traffic_received_bytes counter
minio_s3_traffic_received_bytes{server="127.0.0.1:9000"} 0
# HELP minio_s3_traffic_sent_bytes Total number of s3 bytes sent
# TYPE minio_s3_traffic_sent_bytes counter
minio_s3_traffic_sent_bytes{server="127.0.0.1:9000"} 0
# HELP minio_software_commit_info Git commit hash for the MinIO release.
# TYPE minio_software_commit_info gauge
minio_software_commit_info{commit="DEVELOPMENT.GOGET",server="127.0.0.1:9000"} 0
# HELP minio_software_version_info MinIO Release tag for the server
# TYPE minio_software_version_info gauge
minio_software_version_info{server="127.0.0.1:9000",version="DEVELOPMENT.GOGET"} 0
# HELP promhttp_metric_handler_requests_in_flight Current number of scrapes being served.
# TYPE promhttp_metric_handler_requests_in_flight gauge
promhttp_metric_handler_requests_in_flight 1
# HELP promhttp_metric_handler_requests_total Total number of scrapes by HTTP status code.
# TYPE promhttp_metric_handler_requests_total counter
promhttp_metric_handler_requests_total{code="200"} 1
promhttp_metric_handler_requests_total{code="500"} 0
promhttp_metric_handler_requests_total{code="503"} 0

```

### CODE REVIEW

File: cmd/support-metrics.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+		return e
+	}
+
+	defer resp.Body.Close()
+	respBytes, e := ioutil.ReadAll(io.LimitReader(resp.Body, metricsRespBodyLimit))
+	if e != nil {
+		return e
+	}
+	if resp.StatusCode == http.StatusOK {
+		console.Println(string(respBytes))
```
Comment (harshavardhana) [inline cmd/support-metrics.go:105]: IMO we should pretty this result, instead of showing it in barebones Prometheus format.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/support-metrics.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  {{end}}
+EXAMPLES:
+  1. List of metrics reported cluster wide.
+     {{.Prompt}} {{.HelpName}} play
+
+`,
+}
+
+const (
+	metricsRespBodyLimit = 1 << 20 // 1 MiB
```
[the remark this answers is not in the document]
Comment (sinhaashish) [reply cmd/support-metrics.go]: done

File: cmd/support-metrics.go (hunk 1 above)
Comment (sinhaashish) [reply cmd/support-metrics.go:106]: @harshavardhana  , can you suggest any format to display this result?

Comment (harshavardhana) [reply cmd/support-metrics.go:106]: > @harshavardhana , can you suggest any format to display this result?

First I think is that --json if passed you need to make sure to print these values in JSON format.

This PR is not honoring --json flag. 

Will get back on pretty printing these values...

Comment (sinhaashish) [reply cmd/support-metrics.go]: used `prom2json` to convert prometheus text output to[ json ](https://github.com/prometheus/prom2json)

File: cmd/support-metrics.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	if e != nil {
+		return e
+	}
+
+	req, e := http.NewRequest(http.MethodGet, hostConfig.URL+metricsEndPoint, nil)
+	if e != nil {
+		return e
+	}
+	req.Header.Add("Authorization", "Bearer "+token)
+	resp, e := httpDo(req)
```
Comment (anjalshireesh) [inline cmd/support-metrics.go:98]: This `httpDo` is from `subnet-utils.go` and is probably not the right function to make api calls to minio, as it uses subnet-specific config like the subnet proxy. (Should be renamed to `subnetHTTPDo` I think).

**Review by anjalshireesh — CHANGES_REQUESTED** [review]
As discussed with @abperiasamy and @kannappanr , it might be better to move this command under `mc admin prometheus`

**Comment by sinhaashish** [comment]
> Add a new command `mc support metrics <alias>`

Can take this in a separate PR.

**Comment by sinhaashish** [comment]
> As discussed with @abperiasamy and @kannappanr , it might be better to move this command under `mc admin prometheus`

Raised PR https://github.com/minio/mc/pull/4049 for the same ., thus closing tis PR

> Does the output still look like this? Shouldn't we make it more readable?
> 
> I propose:
> 
> 1. Remove comments by default.
> 2. Format numbers more readable. At least  `3.59258271744e+11` -> `359,258,271,744`. It doesn't have to know units, but scientific notation should go IMO.
> 3. Group by "server", if there can be multiple, or remove it and only print it once.
> 
> ```
> $ ./mc support metrics myminio
> Server 127.0.0.1:9000:
>     minio_cluster_capacity_raw_free_bytes: 1,437,033,086,976
>     minio_cluster_capacity_raw_total_bytes: 1,907,486,965,760
>     minio_cluster_capacity_usable_free_bytes: 18,516,543,488
> ...
> ```

Can take this up in a separate PR

**Comment by sinhaashish** [comment]
> Does the output still look like this? Shouldn't we make it more readable?
> 
> I propose:
> 
> 1. Remove comments by default.
> 2. Format numbers more readable. At least  `3.59258271744e+11` -> `359,258,271,744`. It doesn't have to know units, but scientific notation should go IMO.
> 3. Group by "server", if there can be multiple, or remove it and only print it once.
> 
> ```
> $ ./mc support metrics myminio
> Server 127.0.0.1:9000:
>     minio_cluster_capacity_raw_free_bytes: 1,437,033,086,976
>     minio_cluster_capacity_raw_total_bytes: 1,907,486,965,760
>     minio_cluster_capacity_usable_free_bytes: 18,516,543,488
> ...
> ```

can take this in a seperate PR

**Comment by sinhaashish** [comment]
> As discussed with @abperiasamy and @kannappanr , it might be better to move this command under `mc admin prometheus`

Modified as per the discussion

**Review by harshavardhana — APPROVED** [review]

File: cmd/admin-prometheus-metrics.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+func (pm PrometheusMetrics) String() string {
+	respBytes, e := ioutil.ReadAll(pm.Reader)
+	if e != nil {
+		fatalIf(probe.NewError(e), "error reading metrics:")
+	}
+	return string(respBytes)
+}
+
+// PrometheusMetrics mirrors the MetricFamily proto message.
+type PrometheusMetrics struct {
```
Comment (anjalshireesh) [inline cmd/admin-prometheus-metrics.go:130]:
```suggestion
// prometheusMetrics mirrors the MetricFamily proto message.
type prometheusMetrics struct {
```

Need not be exported as it is local to this file. Also, `prometheusMetricsReader` might be a better name.

**Review by anjalshireesh — COMMENTED** [review]
LGTM, one minor suggestion.

File: cmd/admin-prometheus-metrics.go (hunk 4 above)
Comment (sinhaashish) [reply cmd/admin-prometheus-metrics.go:130]: done

**Review by anjalshireesh — APPROVED** [review]
LGTM

**Comment by harshavardhana** [comment]
PTAL conflicts @sinhaashish

**Pull request merged**